modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxFinishFuture.java

```diff
@@ -1032,6 +1033,25 @@ public GridDistributedTxMapping mapping() {
                                     mini.onDhtFinishResponse(backupId, true);
                             }
                         }
+                        else {
+                            onDone(new TransactionHeuristicException("Primary node [nodeId=" + nodeId + ", consistentId=" +
+                                m.primary().consistentId() + "] has left the grid and there are no backup nodes"));
+
+                            return true;
```

Comment:
Doesn't look good to me. The transaction can still be in prepared state and recovering for other primaries.
I think it's ok to trigger heuristic exception only if near tx is in commiting state and a primary node has failed. No need to check for backups therefore.

Reply:
Test https://ggtc.gridgain.com/project.html?projectId=Tests_GridGainCeEeUe_Latest_CE&testNameId=1912185254194727833&tab=testDetails&branch_Tests_GridGainCeEeUe_Latest_CE=gg-27999 failed when I removed backups check. So I think return it.